### PR TITLE
[#277] Optimizes the JComponent's creation on all inspection classes

### PR DESCRIPTION
**Pull request by rentalhost**
_Okay, it modifies 33 files by a good reason: **readability**.
Future devs will appreciate it a lot._

Following #277, I did this changes to optimize the `JComponent`'s creation on inspections that declares it. It is inevitably part of the inspection class, generated by method `createOptionsPanel()`.

Basically I have create a new class called `OptionsComponent` that can handle tree types of components: _checkboxes_ (via `createCheckbox()`), _radios buttons_ (via `createRadio()`) and _lists_ (via `createList()`).

~~As exception (_included on original code_), the _radio buttons_ is treated like _checkboxes_. It happen only on inspection "_yoda/regular conditions style usage_", and in this case, we need have the possibility to disable this inspection by unchecking both options (_yoda_ and _regular style_). We can create real radio buttons here instead (via `JRadioButton`), but I don't think need and checkbox works for now.~~ Fixed/waiting for #288.

~~Another change that I have did was rename all fields that is treated like an option. Then instead it seems like an static variable (like `SUGGEST_TO_USE_COUNT_CHECK`) now it is treated like an option variable (like `optionSuggestToUseCountCheck`).~~ Reverted to avoid BC (will be discussed later).

In general, it have reduced the code related to inspection options from about 260 lines to 45 (excluding spaces). Which is a reduction of 83%, which means that I will pay less when do a new eye exam. It mean that a lot of duplicate/similar code and needless inner classes was dropped too.

_Well, the next step is do a similar thing with UTs._

### CODE REVIEW

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java [hunk 1]
```diff
@@ -33,8 +34,8 @@
  */
 
 public class UsageOfSilenceOperatorInspector extends BasePhpInspection {
-    // configuration flags automatically saved by IDE
-    public boolean RESPECT_CONTEXT = true;
+    // Inspection options.
+    public boolean optionRespectContext = false;
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: Will destroy custom configurations, names must not change. Know naming is weird, but i meant this naming to spot magically saved options, sorry.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java [hunk 2]
```diff
@@ -79,60 +83,24 @@ public void visitPhpBinaryExpression(BinaryExpression expression) {
...
-            preferRegularStyle.addChangeListener(e -> {
-                PREFER_REGULAR_STYLE = preferRegularStyle.isSelected();
-                if (PREFER_REGULAR_STYLE) {
-                    preferYodaStyle.setSelected(false);
-                    PREFER_YODA_STYLE = false;
-                }
-
-                CONFIGURED = PREFER_YODA_STYLE || PREFER_REGULAR_STYLE;
+        return OptionsComponent.create((component) -> {
+            component.createRadio((radioComponent) -> {
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: Elegant! We need 3rd one: "No preferences" -> CONFIGURED

**Comment by kalessil** [comment]
Elegant, love it.

We need only to recover fields names, as refactoring destroys users settings - nogo. +See 2 other comments.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java (hunk 1 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: Yeah, I have checked that. But there are just few customizable inspections (_about 16 inspections_), then I belive that the user affected will be a small part. And from here to future we can keep the naming convension well.

About _spot magically saved options_, it is just the reason that I prefixed `option`. Then we know that `if(optionX)` is mean to be customizable "X" option. Use a uppercased variable seems that we working with a constant, and make confuses on first step.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: On original code we have only two options. The "configured" was a mix of both. Something like `configured = a.isSelected() || b.isSelected()`. I really think about create a 3rd one to treats a "disabled option", and make more sense too (then we can use a `JRadioButton` instead too to make yet more sense). 

By other side, not make sense user enable the inspection and disable it internally. I guess that we should consider mark some style by default, then user can decide to disable it "as inspection" and not by radio.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java (hunk 1 above)
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: @rentalhost : this is really needed. I need a good analogy, imagine that yo was tweaking IDE code style (e.g. align assignments, + 25 other options you are not remembering anymore and) and PS on update did reset them.

I don't want our community to run into this. The change is a good one, but we can not afford it, pleaseeeeeeeeeeee return naming back.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: Good idea lets keep 2 of them and deactivate the inspection by default.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    public void createRadio(@NotNull final Consumer<RadioComponent> componentDefiner) {
+        final RadioComponent radioComponent = new RadioComponent();
+        componentDefiner.accept(radioComponent);
+    }
+
+    /**
+     * Creates a List.
+     *
+     * @param fieldTitle  The title of this field on panel.
+     * @param listItems   The default or current list items.
```
Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: How it can be `default or current` - this listItems should be always a list of items or i`m wrong?

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+            if (defaultValue) {
+                currentSelection = createElement;
+            }
+        }
+
+        /**
+         * Handle a JCheckBox component and it Update Consumer.
+         */
+        private class Element {
```
Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: Maybe we can rename class to CheckboxElement?

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+             * @param updateConsumer The update consumer (called after item change).
+             */
+            Element(final String fieldTitle, final boolean defaultValue, final Consumer<Boolean> updateConsumer) {
+                this.updateConsumer = updateConsumer;
+
+                checkbox = new JCheckBox(fieldTitle, defaultValue);
+                checkbox.addItemListener((itemEvent) -> {
+                    final boolean isSelected = checkbox.isSelected();
+
+                    if (currentSelection != null && currentSelection != this) {
```
Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: `null!=currentSelection ` ;)

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java (hunk 1 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: Okay, then I will undo that. But, can we reconsider that for 3.x, as a breaking-change? Maybe next year.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: Once opted for this, we should consider that it will have _default checked option_ when this inspection is enabled. I don't think that is need deactivate this inspection by default. Then... could we consider the use the regular style here? The yoda style is more relevant on C language. (Except if you really prefer disable it, but for a big surprise, I don't know about this feature until I work on it, and I was really willing to open an issue to implement it)

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java (hunk 3 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: Initially, I tried to use the `AtomicBoolean` that initially receives the "default value" defined on field of inspection class. Then for all components I create this parameter ("default or current"). Next, I found that it is impossible currently, because IDE don't recognize the `AtomicBoolean` as a configurable option (_it just ignore it_).

The unique type that we can have as a _non-primitive type_ is a `List<>`. Then it could have the default value (which is defined from inspection class) or the current value (after IDE load that). It are "magic fields" which are strangely managed by IDE (rather than doing it more efficiently).

To summarize: this variable is used both to read the default values and to read the current configuration.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java (hunk 4 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: I don't know what is the best name for this case. Note that it is not really a checkbox, not even a radio button. It stores a checkbox and this consumer. Then the `setSelected()` method can both modify the checkbox checked status and trigger the consumer.

But, I do agree that `Element` is not a good name.
New suggestions?

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java (hunk 5 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: Weeeeelll.... I don't agree. In all project code we have a mix of two styles: Yoda and regular.
But let talk about that on this topic: #286.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java (hunk 4 above)
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: Widget?

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: Remembering related discussions, disabling it by default is the most non-conflicting approach. PHP community is pretty aggressive at this specific check (have no idea).

And please make yoda style the default one.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java (hunk 1 above)
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: To do it nicely we need to have transition version with new and old properties and copy over values.
And at some point (6 months later) we can drop old variables.

If you transition version can be done in the PR, we can start with the next release already.

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: Let do panel migration first with the new and nice solution. 2nd PR can be migration.

Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/UsageOfSilenceOperatorInspector.java]: Yes, I do agree that is better implements it first (with original variable names), then care about variable changes in another context (it can be done in another moment, another PR). It because I need see if it is possible currently too (for now I can think about a workaround only, and I like to avoid that).

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: Okay, then it should be disabled by default (the inspection itself), but when enabled we should mark Yoda as default style. Right?

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java (hunk 4 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/options/OptionsComponent.java]: I guess that we can work with `Widget`. Noted!

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java (hunk 2 above)
Comment (rentalhost) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/codeSmell/ComparisonOperandsOrderInspector.java]: In all way, I should open a new issue because it is not related to this PR. #288

**Comment by rentalhost** [comment]
Done all things and rebased to master.

Tests passed successfully.
Tested manually on PS.
No breaking-changes.

Best option to merge (if have no more issues) is squash.

**Comment by kalessil** [comment]
Looks good, will merge today.

**Comment by rentalhost** [comment]
@kalessil Some news here?
I can rebase it to head again before merge it.

**Comment by kalessil** [comment]
Working thru PRs last week, so will be merged soon.

Rebase is a good idea - builds will be triggered then.

**Comment by rentalhost** [comment]
Rebasing done without issues.

**Comment by kalessil** [comment]
Super, I'll take care about merging.

**Comment by kalessil** [comment]
Thank you @rentalhost.
